### PR TITLE
Add unit test for https://github.com/dotnet/msbuild/pull/8870

### PR DESCRIPTION
**Pull request by JaynieBai**
Fixes [8939](https://github.com/dotnet/msbuild/issues/8939)

### Context
https://github.com/dotnet/msbuild/pull/8870#issuecomment-1593945527

### Changes Made
Add a task that returns a custom ITaskItem implementation that has a custom IDictionary type returned from CloneCustomMetadata()
Add unit test TestTaskDictionaryOutputItems()

### CODE REVIEW

File: src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs [hunk 1]
```diff
@@ -1468,4 +1486,32 @@ private ProjectInstance CreateTestProject()
...
+    {
+        public string ItemSpec { get => $"{nameof(DictionaryTaskItem)}spec"; set => throw new NotImplementedException(); }
+
+        public ICollection MetadataNames => throw new NotImplementedException();
+
+        public int MetadataCount => throw new NotImplementedException();
+
+        private Dictionary<string, string> metaData = new() { ["a"] = "b" };
+
+        public IDictionary CloneCustomMetadata() => new Dictionary<string, string>(metaData);
```
Comment (JanKrivanek) [inline src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs:1511]: I believe we need a custom `IDictionary` implementation (not `Dictionary<string, string>`) here to test the https://github.com/dotnet/msbuild/pull/8870 fix

File: src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs [hunk 2]
```diff
@@ -1029,6 +1029,24 @@ public void TestTaskResolutionFailureWithNoUsingTask()
...
         }
 
+        /// <summary>
+        /// https://github.com/dotnet/msbuild/issues/8864
+        /// </summary>
+        [Fact]
+        public void TestTaskDictionaryOutputItems()
+        {
+            string customTaskPath = Assembly.GetExecutingAssembly().Location;
+            MockLogger ml = ObjectModelHelpers.BuildProjectExpectSuccess($"""
```
Comment (JanKrivanek) [inline src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs:1039]: nit: Do not store the return value unless required

File: src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs [hunk 3]
```diff
@@ -1029,6 +1029,24 @@ public void TestTaskResolutionFailureWithNoUsingTask()
             _logger.AssertLogContains("MSB4036");
         }
 
+        /// <summary>
+        /// https://github.com/dotnet/msbuild/issues/8864
+        /// </summary>
+        [Fact]
+        public void TestTaskDictionaryOutputItems()
```
Comment (JanKrivanek) [inline src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs:1036]: Can this somehow test as well that the output has the expected metadata (to verify that the fix aids to properly copying of the metadata)

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Looks as a good starting point.
I believe we need to add custom `IDictionary` implementation and verification of the resulting task output in order to test the fix

**Review by rainersigwald — APPROVED** [review]
LGTM and I believe addresses @JanKrivanek's concerns.

**Review by JanKrivanek — APPROVED** [review]
Thanks!

**Pull request merged**